agent/proxycfg/manager.go

```diff
@@ -0,0 +1,273 @@
+
+	mu       sync.Mutex
+	proxies  map[string]*State
+	watchers map[string]map[uint64]chan *ConfigSnapshot
+}
+
+// NewManager constructs a manager from the provided agent cache.
+func NewManager(logger *log.Logger, cache *cache.Cache, state *local.State,
+	source *structs.QuerySource) *Manager {
+	m := &Manager{
+		logger: logger,
+		cache:  cache,
+		state:  state,
+		source: source,
+		// Single item buffer is enough since there is no data transferred so this
+		// is "level triggering" and we can't miss actual data.
+		stateCh:  make(chan struct{}, 1),
+		proxies:  make(map[string]*State),
+		watchers: make(map[string]map[uint64]chan *ConfigSnapshot),
+	}
+	go m.run()
+	return m
+}
+
+// run is the internal state syncing loop that watches the local state for
+// changes.
+func (m *Manager) run() {
+
+	m.mu.Lock()
+	stateCh := m.stateCh
+	m.mu.Unlock()
+
+	if stateCh == nil {
+		return
+	}
+
+	// Register for notifications about state changes
+	m.state.Notify(stateCh)
+	defer m.state.StopNotify(stateCh)
+
+	for {
+		m.mu.Lock()
+
+		// Traverse the local state and ensure all proxy services are registered
+		services := m.state.Services()
+		for svcID, svc := range services {
+			if svc.Kind != structs.ServiceKindConnectProxy {
+				continue
+			}
+			// TODO(banks): need to work out when to default some stuff. For example
+			// Proxy.LocalServicePort is practially necessary for any sidecar and can
+			// default to the port of the sidecar service, but only if it's already
+			// registered and once we get past here, we don't have enough context to
+			// know that so we'd need to set it here if not during registration of the
+			// proxy service. Sidecar Service and managed proxies in the interim can
+			// do that, but we should validate more generally that that is always
+			// true.
+			err := m.ensureProxyServiceLocked(svc, m.state.ServiceToken(svcID))
+			if err != nil {
+				m.logger.Printf("[ERR] failed to watch proxy service %s: %s", svc.ID,
+					err)
+			}
+		}
+
+		// Now see if any proxies were removed
+		for proxyID := range m.proxies {
+			if _, ok := services[proxyID]; !ok {
+				// Remove them
+				m.removeProxyServiceLocked(proxyID)
+			}
+		}
+
+		m.mu.Unlock()
+
+		// Wait for a state change
+		_, ok := <-stateCh
+		if !ok {
+			// Stopped
+			return
+		}
+	}
+}
+
+// ensureProxyServiceLocked adss or changes the proxy to our state.
+func (m *Manager) ensureProxyServiceLocked(ns *structs.NodeService, token string) error {
+	state, ok := m.proxies[ns.ID]
+
+	if ok {
+		if !state.Changed(ns, token) {
+			// No change
+			return nil
+		}
+
+		// We are updating the proxy, close it's old state
+		state.Close()
+	}
+
+	var err error
+	state, err = NewState(m.logger, m.cache, m.source, ns, token)
+	if err != nil {
+		return err
+	}
+	m.proxies[ns.ID] = state
+
+	// Start a goroutine that will wait for changes and broadcast them to watchers.
+	go func() {
+		ch := state.Watch()
+		for {
+			select {
+			case snap, ok := <-ch:
```

Comment:
If you only have the one case, why bother with the `select` at all? You could just do a bare `snap, ok := <-ch`